### PR TITLE
Add fives_and_sixes example, improve recursion

### PR DESCRIPTION
**Pull request by deosjr**
* removed the need for DefRel, which was a crutch

### CODE REVIEW

File: micro/fivesandsixes_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    want = []*ast.SExpr{ast5, ast5}
+    if !reflect.DeepEqual(got, want) {
+        t.Fatalf("expected %#v but got %#v", got, want)
+    }
+
+	states = RunGoal(
+		10,
+		CallFresh(func(x *ast.SExpr) Goal {
+			return DisjointO(
+				func() GoalFn {return fives(x)},
```
Comment (deosjr) [inline micro/fivesandsixes_test.go:70]: Here you see the difficulty with lexical scope.
How does the ast parser deal with any of that, anyways?

File: micro/fivesandsixes_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
+package micro
+
+import (
+    "reflect"
+    "testing"
+
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+var (
```
Comment (awalterschulze) [inline micro/fivesandsixes_test.go:10]: Seems these can be moved into the test

File: micro/fivesandsixes_test.go (hunk 1 above)
Comment (awalterschulze) [reply micro/fivesandsixes_test.go:70]: I think I forgot enough, that I need a bit more of an explanation and/or context, sorry.

**Review by awalterschulze — APPROVED** [review]
The cleanup looks nice and new test seems great, but maybe I am misinterpreting something.

File: micro/fivesandsixes_test.go (hunk 1 above)
Comment (deosjr) [reply micro/fivesandsixes_test.go:70]: So in my first attempt at this, I wasnt passing x into fives at all (see the first testcase).
Which works in this case. In the general case though, we want to be able to compose defined functions.
Right now if you would name the var x in fives and y in sixes in that way, you wouldn't get the correct answer.
Solution for that is to pass the callfresh var into the fives func like I am doing here, but that needs the boilerplate you see here in order to work. So clearly that is not ideal.

I would like to be able to define fives in terms of x, and have that x be scoped only for the func, without having to pass the var into the function from the bigger func. That would need dealing with scope, which we currently aren't doing at all. This example is from the paper, I'm using it here to demonstrate the scope issue.
Did that make it more clear at all?

Comment (awalterschulze) [reply micro/fivesandsixes_test.go]: Your pull request made finally made it clear to me.
https://github.com/awalterschulze/gominikanren/pull/9
Sorry I was a bit slow on understanding the issue, but thanks for fixing it <3

**Pull request merged**